lib/python/treadmill_aws/ipa525.py

```diff
@@ -0,0 +1,217 @@
+"""Implements IPA525 credential service.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import base64
+import grp
+import io
+import logging
+import json
+import os
+import pwd
+import re
+import shutil
+import tempfile
+
+from twisted.internet import reactor
+from twisted.internet import protocol
+
+from treadmill import gssapiprotocol
+from treadmill import utils
+
+_LOGGER = logging.getLogger(__name__)
+
+
+# Disable warning for too many branches.
+# pylint: disable=R0912
+def run_server(port, client_realm, server_realm, admin_group):
+    """Runs IPA525 server."""
+    _LOGGER.info(
+        'IPA 525 server starting - listening on port %d',
+        port)
+
+    def _parse_name(princ):
+        match = re.search('^([^/@]*)/?([^@]*)@?(.*)$', princ)
+        if match is None:
+            _LOGGER.error(
+                'Internal error - expected name to be a kerberos '
+                'principal name, got [%s]', princ)
+            raise ValueError('Internal error')
+
+        name = match.group(1)
+        inst = match.group(2)
+        realm = match.group(3)
+
+        return [name, inst, realm]
+
+    class IPA525Server(gssapiprotocol.GSSAPILineServer):
+        """IPA525 server."""
+
+        def __init__(self, client_realm, server_realm, admin_group):
+            gssapiprotocol.GSSAPILineServer.__init__(self)
+            self.client_realm = client_realm
+            self.server_realm = server_realm
+            self.admin_group = admin_group
+
+        def _validate_request(self, request):
+            name, inst, realm = _parse_name(request)
+
+            if inst:
+                raise ValueError(
+                    'Request [%s] is for a service principal - not supported' %
+                    request)
+
+            if realm:
+                if realm != self.server_realm:
+                    raise ValueError(
+                        'Request [%s] is for principal in an unsupported '
+                        'service realm [%s]' % (request, realm))
+
+            try:
+                uid = pwd.getpwnam(name)
+            except KeyError:
+                raise KeyError("%s is not a valid username" % name)
+
+        def _authorize(self, requestor, request):
+
+            if self._authorize_admin(requestor):
+                return
+
+            self._authorize_self(requestor, request)
+
+        def _authorize_admin(self, requestor):
```

Comment:
nit - it took me while to understand that _auth_admin return bool and throws, and _auth_self just throws. 

i suggest you make _auth_self return bool (for consistency), add request to _auth_admins, and then you can have:

_authorizers = [
  self._auth_admin,
  self._auth_self
]

_authorisze(user, request):
  for _auth in self._authorizers:
     if _auth(user, request): 
        return True

return False

You may consider a base class for all this, as this is common functionality for all three servers.